### PR TITLE
Added database fragmentation check

### PR DESCRIPTION
**Pull request by shivanshuag**
Issue #51 
To test,

``` bash
 ❯ drush @d8 ad --detail                                [13:57:58]
Database: 83%
  Total size: Determine the size of the database.
    Total size: 7.41MB
  Collations: Check to see if there are any tables that aren't using UTF-8.
    Every table is using UTF-8.
  Storage Engines: Check to see if there are any tables that aren't using InnoDB.
    Every table is using InnoDB.
  Tables with at least 1000 rows: Return list of all tables with at least 1000 rows in the database.
    No tables with less than 1000 rows.
  Database Fragmentation: Reports all the tables with fragmentation greater than 5%
    Table Name: Fragmentation Percentage
    ---------------------
    cache_discovery: 400
      Run "OPTIMIZE TABLE" on the fragmented tables. Refer to https://dev.mysql.com/doc/refman/5.5/en/optimize-table.html for more details.
```

``` bash
 ❯ mysql -u root -p                                     [13:58:05]
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 128
Server version: 10.0.20-MariaDB-log MariaDB Server

Copyright (c) 2000, 2015, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> OPTIMIZE TABLE drupal8.cache_discovery;
+-------------------------+----------+----------+-------------------------------------------------------------------+
| Table                   | Op       | Msg_type | Msg_text                                                          |
+-------------------------+----------+----------+-------------------------------------------------------------------+
| drupal8.cache_discovery | optimize | note     | Table does not support optimize, doing recreate + analyze instead |
| drupal8.cache_discovery | optimize | status   | OK                                                                |
+-------------------------+----------+----------+-------------------------------------------------------------------+
2 rows in set (2.63 sec)

MariaDB [(none)]> exit
Bye
```

``` bash
 ❯ drush @d8 ad --detail                                [13:58:29]
Database: 100%
  No action required.                                   [success]
  Total size: Determine the size of the database.
    Total size: 7.55MB
  Collations: Check to see if there are any tables that aren't using UTF-8.
    Every table is using UTF-8.
  Storage Engines: Check to see if there are any tables that aren't using InnoDB.
    Every table is using InnoDB.
  Tables with at least 1000 rows: Return list of all tables with at least 1000 rows in the database.
    No tables with less than 1000 rows.
  Database Fragmentation: Reports all the tables with fragmentation greater than 5%
    None of the tables has fragmentation of more than 5%
```

### CODE REVIEW

File: Check/Database/Fragmentation.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    return $ret_val;
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getAction().
+   */
+  public function getAction() {
+    if ($this->getScore() == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_WARN) {
+      return dt('Run "OPTIMIZE TABLE" on the fragmented tables. Refer to https://dev.mysql.com/doc/refman/5.5/en/optimize-table.html for more details.');
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Use https://dev.mysql.com/doc/en/optimize-table.html as it's evergreen and will automatically redirect to current stable release.

File: Check/Database/Fragmentation.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  public function getLabel() {
+    return dt('Database Fragmentation');
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getDescription().
+   */
+  public function getDescription() {
+    return dt("Reports all the tables with fragmentation greater than 5%");
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Single, not double quotes.

`Detect table fragmentation which increases storage space and decreases I/O efficiency.`

File: Check/Database/Fragmentation.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      $sql = drush_sql_get_class();
+      $db_spec = $sql->db_spec();
+    }
+    else {
+      $db_spec = _drush_sql_get_db_spec();
+    }
+    $sql_query  = 'SELECT TABLE_NAME AS name ';
+    $sql_query .= ', Round(DATA_LENGTH/1024/1024) as data_length ';
+    $sql_query .= ', Round(INDEX_LENGTH/1024/1024) as index_length ';
+    $sql_query .= ', Round(DATA_FREE/ 1024/1024) as data_free ';
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Same operator spacing as PHP.

Use capitalization on verbs - https://coderwall.com/p/f_anjq/best-practices-for-writing-procedural-sql-queries?p=1&q=author%3Afluxsauce

**Comment by fluxsauce** [comment]
> cache_discovery: 400

400% fragmentation? Are you sure about that?

**Comment by shivanshuag** [comment]
I am not sure what would percentage of fragmentation be. I have changed it to `free_space/(used_space + free_space)`. Is it all right?

File: Check/Database/Fragmentation.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    $sql_query .= 'FROM information_schema.TABLES ';
+    $sql_query .= 'WHERE TABLES.DATA_FREE > 0 ';
+    $sql_query .= 'AND TABLES.table_schema = :dbname ';
+    $result = db_query($sql_query, array(
+      ':dbname' => $db_spec['database'],
+    ));
+    foreach ($result as $row) {
+      $data = $row->data_length + $row->index_length;
+      $free = $row->data_free;
+      $fragmentation_ratio = ($free / ($data + $free)) * 100;
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: free / data

File: Check/Database/Fragmentation.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultWarn().
+   */
+  public function getResultWarn() {
+    arsort($this->registry['database_fragmentation']);
+    if (drush_get_option('html')) {
+      $ret_val = '<table class="table table-condensed">';
+      $ret_val .= '<thead><tr><th>' . dt('Table Name') . '</th><th>' . dt('Fragmentation Percentage') . '</th></tr></thead>';
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Fragmentation ratio

File: Check/Database/Fragmentation.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        $ret_val .= '<tr>';
+        $ret_val .= '<td>' . $name . '</td>';
+        $ret_val .= '<td>' . $ratio . '</td>';
+        $ret_val .= '</tr>';
+      }
+      $ret_val .= '</tbody>';
+      $ret_val .= '</table>';
+    }
+    else {
+      $ret_val  = dt('Table Name: Fragmentation Percentage') . PHP_EOL;
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Fragmentation ratio

File: Check/Database/Fragmentation.php [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  public function calculateScore() {
+    if (version_compare(DRUSH_VERSION, 7, '>=')) {
+      $sql = drush_sql_get_class();
+      $db_spec = $sql->db_spec();
+    }
+    else {
+      $db_spec = _drush_sql_get_db_spec();
+    }
+    $sql_query  = 'SELECT TABLE_NAME AS name';
+    $sql_query .= ', Round(DATA_LENGTH/1024/1024) AS data_length';
```
Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: Capitalize verbs and spacing around variables still.

Comment (shivanshuag) [inline Check/Database/Fragmentation.php]: I am not sure which verb your are referring to. `AS` is in capitals. I also don't see what is wrong with the spacing. Please help me out here.

Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: A verb is a word used to describe an action, state, or occurrence.

`ROUND`

Comment (fluxsauce) [inline Check/Database/Fragmentation.php]: > also don't see what is wrong with the spacing.

`(DATA_LENGTH / 1024 / 1024)`

Same as PHP.

**Comment by fluxsauce** [comment]
```
jon@Decker ~/projects/drupal-8.0.0-beta10 $ drush ad
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Division by zero Fragmentation.php:110                                                                                                             [warning]
Database: 75%
  Database Fragmentation
    Table Name: Fragmentation Ratio                                                                                                                [warning]
    ---------------------
    key_value: 69
    router: 34.5
      Run "OPTIMIZE TABLE" on the fragmented tables. Refer to https://dev.mysql.com/doc/en/optimize-table.html for more details.
```

**Comment by fluxsauce** [comment]
@shivanshuag - reminder, please close and re-submit this with a clean branch.

**Comment by shivanshuag** [comment]
@fluxsauce If I remove the other commits(pagespeed v2 and folder structure), they will be lost, because I do not see them in 8.x-2.x branch.

**Comment by fluxsauce** [comment]
@shivanshuag oh no, really? What the heck? Okay. Resolve the merge conflict, please.

**Comment by shivanshuag** [comment]
@fluxsauce fixed the conflicts.

**Pull request merged**